File: 2024/day17/perl/solution.pl
#!/usr/bin/env perl
# Day 17: Chronospatial Computer - 3-bit VM emulator
use strict;
use warnings;
use feature 'say';
use FindBin qw($RealBin);

# Parse input file
sub parse_input {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    my $text = do { local $/; <$fh> };
    close $fh;

    my ($a) = $text =~ /Register A: (\d+)/;
    my ($b) = $text =~ /Register B: (\d+)/;
    my ($c) = $text =~ /Register C: (\d+)/;
    my ($prog_str) = $text =~ /Program: ([\d,]+)/;
    my @program = split /,/, $prog_str;

    return ($a, $b, $c, \@program);
}

# Execute the 3-bit computer program and return output
sub run_program {
    my ($a, $b, $c, $program) = @_;
    my $ip = 0;
    my @output;

    # Get combo operand value
    my $combo = sub {
        my ($operand) = @_;
        return $operand if $operand <= 3;
        return $a if $operand == 4;
        return $b if $operand == 5;
        return $c if $operand == 6;
        die "Invalid combo operand: $operand";
    };

    while ($ip < @$program) {
        my $opcode = $program->[$ip];
        my $operand = $program->[$ip + 1];

        if ($opcode == 0) {         # adv - A = A >> combo
            $a = $a >> $combo->($operand);
        }
        elsif ($opcode == 1) {      # bxl - B = B XOR literal
            $b = $b ^ $operand;
        }
        elsif ($opcode == 2) {      # bst - B = combo % 8
            $b = $combo->($operand) & 7;
        }
        elsif ($opcode == 3) {      # jnz - jump if A != 0
            if ($a != 0) {
                $ip = $operand;
                next;
            }
        }
        elsif ($opcode == 4) {      # bxc - B = B XOR C
            $b = $b ^ $c;
        }
        elsif ($opcode == 5) {      # out - output combo % 8
            push @output, $combo->($operand) & 7;
        }
        elsif ($opcode == 6) {      # bdv - B = A >> combo
            $b = $a >> $combo->($operand);
        }
        elsif ($opcode == 7) {      # cdv - C = A >> combo
            $c = $a >> $combo->($operand);
        }

        $ip += 2;
    }

    return \@output;
}

# Part 1: Run the program and return comma-separated output
sub part1 {
    my ($a, $b, $c, $program) = @_;
    my $output = run_program($a, $b, $c, $program);
    return join(',', @$output);
}

# Part 2: Find initial A value that makes program output itself
sub part2 {
    my ($a, $b, $c, $program) = @_;
    my @prog = @$program;
    my $prog_len = @prog;

    # Recursive search - work backwards from the last digit
    # Build A 3 bits at a time
    # Note: We declare $search first, then assign the sub, to enable recursion
    # within the closure (the sub references $search which must exist)
    my $search;
    $search = sub {
        my ($target_idx, $current_a) = @_;

        # Found complete solution
        return $current_a if $target_idx < 0;

        # Try all 8 possible 3-bit values for this position
        for my $bits (0..7) {
            my $candidate_a = ($current_a << 3) | $bits;

            # A can't be 0 at start (would halt immediately)
            next if $candidate_a == 0 && $target_idx == $prog_len - 1;

            my $output = run_program($candidate_a, $b, $c, $program);

            # Check if output matches the suffix of the program
            my @expected = @prog[$target_idx .. $#prog];

            if ("@$output" eq "@expected") {
                my $result = $search->($target_idx - 1, $candidate_a);
                return $result if defined $result;
            }
        }

        return;
    };

    return $search->($prog_len - 1, 0);
}

# Main
my $input_file = "$RealBin/../input.txt";

my ($a, $b, $c, $program) = parse_input($input_file);

say "Part 1: ", part1($a, $b, $c, $program);
say "Part 2: ", part2($a, $b, $c, $program);

File: 2024/day17/clojure/solution.clj
#!/usr/bin/env clojure -M

;; Day 17: Chronospatial Computer - 3-bit VM emulator

(ns solution
  (:require [clojure.string :as str]))

(defn parse-input
  "Parse registers and program from input."
  [text]
  (let [[line-a line-b line-c _ line-prog] (str/split-lines (str/trim text))
        a (parse-long (second (re-find #": (\d+)" line-a)))
        b (parse-long (second (re-find #": (\d+)" line-b)))
        c (parse-long (second (re-find #": (\d+)" line-c)))
        program-str (second (re-find #"Program: ([\d,]+)" line-prog))
        program (vec (map parse-long (str/split program-str #",")))]
    {:a a :b b :c c :program program}))

(defn combo-operand
  "Get combo operand value."
  [operand a b c]
  (case operand
    0 0
    1 1
    2 2
    3 3
    4 a
    5 b
    6 c
    (throw (ex-info "Invalid combo operand" {:operand operand}))))

(defn run-program
  "Execute the 3-bit computer program and return output."
  [a b c program]
  (let [program-len (count program)]
    (loop [ip 0
           a a
           b b
           c c
           output []]
      (if (>= ip program-len)
        output
        (let [opcode (nth program ip)
              operand (nth program (inc ip))]
          (case opcode
            ;; adv - A = A >> combo
            0 (let [shift (combo-operand operand a b c)]
                (recur (+ ip 2) (bit-shift-right a shift) b c output))

            ;; bxl - B = B XOR literal
            1 (recur (+ ip 2) a (bit-xor b operand) c output)

            ;; bst - B = combo % 8
            2 (let [val (bit-and (combo-operand operand a b c) 7)]
                (recur (+ ip 2) a val c output))

            ;; jnz - jump if A != 0
            3 (if (not= a 0)
                (recur operand a b c output)
                (recur (+ ip 2) a b c output))

            ;; bxc - B = B XOR C (ignores operand)
            4 (recur (+ ip 2) a (bit-xor b c) c output)

            ;; out - output combo % 8
            5 (let [val (bit-and (combo-operand operand a b c) 7)]
                (recur (+ ip 2) a b c (conj output val)))

            ;; bdv - B = A >> combo
            6 (let [shift (combo-operand operand a b c)]
                (recur (+ ip 2) a (bit-shift-right a shift) c output))

            ;; cdv - C = A >> combo
            7 (let [shift (combo-operand operand a b c)]
                (recur (+ ip 2) a b (bit-shift-right a shift) output))))))))

(defn part1
  "Run the program and return comma-separated output."
  [text]
  (let [{:keys [a b c program]} (parse-input text)
        output (run-program a b c program)]
    (str/join "," output)))

(defn part2
  "Find initial A value that makes program output itself."
  [text]
  (let [{:keys [b c program]} (parse-input text)
        program-len (count program)]

    ;; Work backwards from the last digit - build A 3 bits at a time
    (letfn [(search [target-idx current-a]
              (if (< target-idx 0)
                current-a
                ;; Try all 8 possible 3-bit values for this position
                (some (fn [bits]
                        (let [candidate-a (bit-or (bit-shift-left current-a 3) bits)]
                          ;; A can't be 0 at start (would halt immediately without output)
                          (when-not (and (= candidate-a 0) (= target-idx (dec program-len)))
                            (let [output (run-program candidate-a b c program)
                                  expected (subvec program target-idx)]
                              (when (= output expected)
                                (search (dec target-idx) candidate-a))))))
                      (range 8))))]
      (search (dec program-len) 0))))

(defn -main []
  (let [text (slurp "../input.txt")]
    (println "Part 1:" (part1 text))
    (println "Part 2:" (part2 text))))

(-main)
